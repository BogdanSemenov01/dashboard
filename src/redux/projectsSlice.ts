import { PayloadAction, createSlice } from '@reduxjs/toolkit'

type State = {
  projects: Array<Project>
  currentProjectId: number
}

export type Project = {
  title: string
  id: number
  tasks: Tasks

}


type Tasks = {
  queueTasks: StatusTasks
  developmentTasks: StatusTasks
  doneTasks: StatusTasks
}

type StatusTasks = Array<Task>


type Task = {
  id: number
  text: string
  description: string
  priority: 'low' | 'middle' | 'high'
  subTasks: Array<Subtask>
}

type Subtask = {
  id: number
  text: string
  isComplete: boolean
}


const initialState: State = {
  projects: [
    {
      title: 'DashBoard',
      id: 1,
      tasks: {
        queueTasks: [
          {
            id: 1,
            text: 'queue',
            description: '',
            priority: 'low',
            subTasks: [
              { id: 1, text: '1', isComplete: true },
              { id: 2, text: '2', isComplete: true },
              { id: 3, text: '3', isComplete: false },
              { id: 4, text: '4', isComplete: false },
            ]
          },
        ],
        developmentTasks: [
          {
            id: 2, 
            text: 'dev', 
            description: '', 
            priority: 'low', 
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
          },
        ],
        doneTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
      }
    },
    {
      title: 'Portfolio',
      id: 2,
      tasks: {
        queueTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
        developmentTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
        doneTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
      }
    }
  ],
  currentProjectId: 0,
}

export const projectsSlice = createSlice({
  name: 'projects',
  initialState,
  reducers: {
    createProject: (state, action: PayloadAction<{text: string}>) => {
      state.projects.push({
        title: action.payload.text,
        id: Date.now(),
        tasks: {
          queueTasks: [],
          developmentTasks: [],
          doneTasks: [],
        }
      })
    },
    deleteProject: (state, action: PayloadAction<{id: number}>) => {
      const index = state.projects.findIndex((p: Project) => p.id === +action.payload.id)
      if (index > -1) {
        state.projects.splice(index, 1)
      }
    },
    setCurrentProjectId: (state, action: PayloadAction<{projectId: string | undefined}>) => {
      state.currentProjectId = Number(action.payload.projectId)
    },
    changeTaskStatus: (state, action: PayloadAction<{
      currentStatus: string
      id: number
      nextStatus?: string
    }>) => {
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          const element = removeTaskFromCurrentStatus(p.tasks, action.payload.currentStatus, action.payload.id)
          pushTaskToNextStatus(p.tasks, element, action.payload.nextStatus)
        }
      })
    },
    createNewTask: (state, action: PayloadAction<{
      status: string
      text: string

    }>) => {
      let statusTask = selectSection(action.payload.status)
      let element: Task = {
        id: Date.now(),
        text: action.payload.text || 'Update me',
        description: '',
        priority: 'low',
        subTasks: []
      }
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          p.tasks[statusTask].push(element)
        }
      })
    },
    deleteTask: (state, action: PayloadAction<{
      section: string
      taskId: number
    }>) => {
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          let section = selectSection(action.payload.section)
          let index = p.tasks[section].findIndex((t: Task) => t.id === action.payload.taskId)
          if (index > -1) {
            p.tasks[section].splice(index, 1)
          }
        }
      })
    },
    changeTask: (state, action: PayloadAction<{
      section: string
      taskId: number
      newText: string
      newDescription: string
      newPriority: "low" | "middle" | "high"
      newSubTasks: Array<Subtask>
    }>) => {
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          let section = selectSection(action.payload.section)
          p.tasks[section].map((t: Task) => {
            if (t.id == action.payload.taskId) {
              t.text = action.payload.newText
              t.description = action.payload.newDescription
              t.priority = action.payload.newPriority
              t.subTasks = [...action.payload.newSubTasks]
            }
          })
        }
      })
    }
  },
})

export const {
  createProject,
  deleteProject,
  setCurrentProjectId,
  changeTaskStatus,
  createNewTask,
  deleteTask,
  changeTask,
} = projectsSlice.actions

export default projectsSlice.reducer


const pushTaskToNextStatus = (state: Tasks, element: Task, nextStatus: string | undefined): void => {
  let section = selectSection(nextStatus)
  state[section].push(element)
}

const removeTaskFromCurrentStatus = (state: Tasks, currentStatus: string, id: number): Task => {
  let section = selectSection(currentStatus)
  const index = state[section].findIndex((el: Task) => el.id === id)
  const result = state[section].splice(index, 1)[0]
  return result
}



const selectSection = (status: string | undefined): 'queueTasks' | 'developmentTasks' | 'doneTasks' => {
  switch (status) {
    case 'Queue':
      return 'queueTasks'
    case 'Development':
      return 'developmentTasks'
    case 'Done':
      return 'doneTasks'
    default:
      return 'queueTasks'
  }
}