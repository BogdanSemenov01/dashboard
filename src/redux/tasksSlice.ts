import {createSlice} from '@reduxjs/toolkit'
import type {PayloadAction} from '@reduxjs/toolkit'

const initialState:any = {
  tasks: {

  }
}

const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    loadTasks: (state, action) => {
      state.tasks.push(action.payload.tasks)
    },
    changeTaskStatus: (state, action) => {
      if (state.tasks) {

        state.tasks.map((t:any) => {
          if (t.id === action.payload.id) {
            t.status = action.payload.status
          }
        })
      }
    }
  },
})


