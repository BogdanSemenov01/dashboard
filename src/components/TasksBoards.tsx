import React, { FC, useEffect } from 'react'
import Section from './Section'
import Flex from './styled/CommonStyledComponents'
import { RootState, useAppDispatch, useAppSelector } from '../redux/store'
import { useParams } from 'react-router-dom'
import Task from './Task'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import { Project, setCurrentProjectId } from '../redux/projectsSlice'

const blueCS = {
  border: 'rgb(38, 131, 182)',
  background: 'rgba(45, 138, 186, 0.2)',
  title: 'rgb(38, 131, 182)'
}
const orangeCS = {
  border: 'rgb(255, 121, 25)',
  background: 'rgba(255, 123, 29, 0.212)',
  title: 'rgb(255, 121, 25)',
}
const greenCS = {
  border: 'rgb(29, 182, 70)',
  background: 'rgba(50, 202, 90, 0.2)',
  title: 'rgb(29, 182, 70)'
}


const TasksBoards: FC = () => {
  const projects = useAppSelector((state: RootState) => state.projects.projects)
  const dispatch = useAppDispatch()
  const projectId = useParams().id
  useEffect(() => {
    dispatch(setCurrentProjectId({projectId}))
  }, [])
  
  const currentProject: Array<any> = [] 
  projects.map((p:Project) => {
    if(p.id.toString() === projectId) {
      currentProject.push(p)
    }
  })


  const tasks = currentProject[0].tasks
  const queueTasks = tasks.queueTasks
  const developmentTasks = tasks.developmentTasks
  const doneTasks = tasks.doneTasks

  const renderTasks = (sectionTitle:string, tasks: []) => {
    if (tasks) {
      return tasks
      .map((t:any, index:any) => {
        return (
          <Task 
            id={t.id} 
            text={t.text} 
            key={t.id} 
            index={index} 
            status={sectionTitle}
            description={t.description}
            priority={t.priority}
            subTasks={t.subTasks}
            />
        )
      })
    }
  }


  return (
    <div>
      <DndProvider backend={HTML5Backend}>
        <Flex justifyContent='space-between' gap='10px'>
          <Section title={'Queue'} colorSC={blueCS}>
            {renderTasks('Queue', queueTasks)}
          </Section>
          <Section title={'Development'} colorSC={orangeCS}>
            {renderTasks('Development', developmentTasks)}
          </Section>
          <Section title={'Done'} colorSC={greenCS}>
            {renderTasks('Done', doneTasks)}
          </Section>
        </Flex>
      </DndProvider>
    </div>
  )
}

export default TasksBoards