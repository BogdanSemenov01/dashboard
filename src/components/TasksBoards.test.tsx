import { render, screen } from "@testing-library/react"
import TasksBoards from "./TasksBoards"
import * as appHooks from "../redux/store"
import { Project } from "../redux/projectsSlice"
import * as useParams from "react-router-dom"

jest.mock("../redux/store")
jest.mock("react-router-dom")

const projects: Array<Project> = [
  {
    title: "DashBoard",
    id: 1,
    tasks: {
      queueTasks: [
        {
          id: 1,
          text: "any",
          description: "any",
          priority: "low",
          subTasks: [],
        },
      ],
      developmentTasks: [
        {
          id: 2,
          text: "any",
          description: "any",
          priority: "low",
          subTasks: [],
        },
      ],
      doneTasks: [
        {
          id: 3,
          text: "any",
          description: "any",
          priority: "low",
          subTasks: [],
        },
      ],
    },
  },
]

const mockedDispatch = jest.spyOn(appHooks, "useAppDispatch")
const mockedSelector = jest.spyOn(appHooks, "useAppSelector")
const mockedParams = jest.spyOn(useParams, "useParams")

describe("Task Board", () => {
  it("should render task board component", () => {
    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)
    mockedSelector.mockReturnValue(projects)
    mockedParams.mockReturnValue({ id: "1" })
    const component = render(<TasksBoards />)
    expect(component).toMatchSnapshot()
    expect(dispatch).toHaveBeenCalledTimes(1)
  })
})
