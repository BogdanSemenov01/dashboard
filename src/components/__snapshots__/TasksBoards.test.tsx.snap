// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Board should render task board component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="sc-bcXHqe hVCnkf"
        >
          <div
            class="sc-hLBbgP bNpxGq"
          >
            <div
              class="sc-bcXHqe auLWH"
            >
              <div
                class="sc-eDvSVe jsUpMt"
              >
                Queue
              </div>
              <div
                class="sc-jrcTuL hBqAWt"
                draggable="true"
                style="opacity: 1;"
              >
                <div
                  class="sc-ipEyDJ jNZfST"
                  result="NaN"
                >
                  <span />
                </div>
                <div>
                  <p
                    class="sc-kDvujY hBxPJI"
                    color="green"
                  >
                    !
                  </p>
                  any
                </div>
                <div
                  class="controls"
                >
                  <button>
                    s
                  </button>
                  <button
                    class="sc-jSUZER iXDapD"
                  >
                    x
                  </button>
                </div>
              </div>
              <button
                class="sc-gswNZR dGHIbY"
              >
                +
              </button>
            </div>
          </div>
          <div
            class="sc-hLBbgP fRuQZs"
          >
            <div
              class="sc-bcXHqe auLWH"
            >
              <div
                class="sc-eDvSVe jsUpMt"
              >
                Development
              </div>
              <div
                class="sc-jrcTuL koTQWt"
                draggable="true"
                style="opacity: 1;"
              >
                <div
                  class="sc-ipEyDJ jNZfST"
                  result="NaN"
                >
                  <span />
                </div>
                <div>
                  <p
                    class="sc-kDvujY hBxPJI"
                    color="green"
                  >
                    !
                  </p>
                  any
                </div>
                <div
                  class="controls"
                >
                  <button>
                    s
                  </button>
                  <button
                    class="sc-jSUZER iXDapD"
                  >
                    x
                  </button>
                </div>
              </div>
              <button
                class="sc-gswNZR ioYtda"
              >
                +
              </button>
            </div>
          </div>
          <div
            class="sc-hLBbgP WNvMq"
          >
            <div
              class="sc-bcXHqe auLWH"
            >
              <div
                class="sc-eDvSVe jsUpMt"
              >
                Done
              </div>
              <div
                class="sc-jrcTuL dOFvOd"
                draggable="true"
                style="opacity: 1;"
              >
                <div
                  class="sc-ipEyDJ jNZfST"
                  result="NaN"
                >
                  <span />
                </div>
                <div>
                  <p
                    class="sc-kDvujY hBxPJI"
                    color="green"
                  >
                    !
                  </p>
                  any
                </div>
                <div
                  class="controls"
                >
                  <button>
                    s
                  </button>
                  <button
                    class="sc-jSUZER iXDapD"
                  >
                    x
                  </button>
                </div>
              </div>
              <button
                class="sc-gswNZR lozcVE"
              >
                +
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="sc-bcXHqe hVCnkf"
      >
        <div
          class="sc-hLBbgP bNpxGq"
        >
          <div
            class="sc-bcXHqe auLWH"
          >
            <div
              class="sc-eDvSVe jsUpMt"
            >
              Queue
            </div>
            <div
              class="sc-jrcTuL hBqAWt"
              draggable="true"
              style="opacity: 1;"
            >
              <div
                class="sc-ipEyDJ jNZfST"
                result="NaN"
              >
                <span />
              </div>
              <div>
                <p
                  class="sc-kDvujY hBxPJI"
                  color="green"
                >
                  !
                </p>
                any
              </div>
              <div
                class="controls"
              >
                <button>
                  s
                </button>
                <button
                  class="sc-jSUZER iXDapD"
                >
                  x
                </button>
              </div>
            </div>
            <button
              class="sc-gswNZR dGHIbY"
            >
              +
            </button>
          </div>
        </div>
        <div
          class="sc-hLBbgP fRuQZs"
        >
          <div
            class="sc-bcXHqe auLWH"
          >
            <div
              class="sc-eDvSVe jsUpMt"
            >
              Development
            </div>
            <div
              class="sc-jrcTuL koTQWt"
              draggable="true"
              style="opacity: 1;"
            >
              <div
                class="sc-ipEyDJ jNZfST"
                result="NaN"
              >
                <span />
              </div>
              <div>
                <p
                  class="sc-kDvujY hBxPJI"
                  color="green"
                >
                  !
                </p>
                any
              </div>
              <div
                class="controls"
              >
                <button>
                  s
                </button>
                <button
                  class="sc-jSUZER iXDapD"
                >
                  x
                </button>
              </div>
            </div>
            <button
              class="sc-gswNZR ioYtda"
            >
              +
            </button>
          </div>
        </div>
        <div
          class="sc-hLBbgP WNvMq"
        >
          <div
            class="sc-bcXHqe auLWH"
          >
            <div
              class="sc-eDvSVe jsUpMt"
            >
              Done
            </div>
            <div
              class="sc-jrcTuL dOFvOd"
              draggable="true"
              style="opacity: 1;"
            >
              <div
                class="sc-ipEyDJ jNZfST"
                result="NaN"
              >
                <span />
              </div>
              <div>
                <p
                  class="sc-kDvujY hBxPJI"
                  color="green"
                >
                  !
                </p>
                any
              </div>
              <div
                class="controls"
              >
                <button>
                  s
                </button>
                <button
                  class="sc-jSUZER iXDapD"
                >
                  x
                </button>
              </div>
            </div>
            <button
              class="sc-gswNZR lozcVE"
            >
              +
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
